eval/bundled/epm.elv.go

```diff
@@ -1,111 +1,279 @@
+        -domain-config[$dom] = $-default-domain-config[$dom]
+        -debug "No existing config for "$dom", using the default: "(to-string $-domain-config[$dom])
+        mkdir -p (dirname $cfgfile)
+        put $-domain-config[$dom] | to-json > $cfgfile
+      } else {
+        fail "No existing config for "$dom" and no default available. Please create config file "(tilde-abbr $cfgfile)" by hand"
+      }
     }
+  }
 }
 
-fn -uninstall-one [pkg]{
-    installed-pkgs = [(installed)]
-    if (not (has-value $installed-pkgs $pkg)) {
-        -error 'Package '$pkg' is not registered as installed.'
-        return
+# Return the method by which a package is installed
+fn -package-method [pkg]{
+  dom = (-package-domain $pkg)
+  -read-domain-config $dom
+  if (has-key $-domain-config $dom) {
+    cfg = $-domain-config[$dom]
+    put $cfg[method]
+  } else {
+    put $false
+  }
+}
+
+# Invoke package operations defined in $-method-handler above
+fn -package-op [pkg what]{
+  dom = (-package-domain $pkg)
+  -read-domain-config $dom
+  if (has-key $-domain-config $dom) {
+    cfg = $-domain-config[$dom]
+    method = $cfg[method]
+    if (has-key $-method-handler $method) {
+      if (has-key $-method-handler[$method] $what) {
+        $-method-handler[$method][$what] $pkg $cfg
+      } else {
+        fail "No handler for '"$what"' defined in method '"$method"'"
+      }
+    } else {
+      fail "No handler defined for method '"$method"', specified in in config file "(-domain-config-file $dom)
     }
-    dest = $-data-dir/lib/$pkg
-    if (not ?(test -d $dest)) {
-        -error 'Package '$pkg' does not exist locally.'
-        return
+  }
+}
+
+# Read and parse the package metadata, if it exists
+fn metadata [pkg]{
+  res = [&]
+  mdata = (-package-metadata-file $pkg)
+  if (and (is-installed $pkg) ?(test -f $mdata)) {
+    res = (cat $mdata | from-json)
+  }
+  put $res
+}
+
+# Print out information about a package
+fn query [pkg]{
+  data = (metadata $pkg)
+  echo (edit:styled "Package "$pkg cyan)
+  echo (edit:styled "Source:" blue) (-package-method $pkg) (-package-op $pkg src)
+  if (is-installed $pkg) {
+    echo (edit:styled "Installed at "(dest $pkg) green)
+  } else {
+    echo (edit:styled "Not installed" red)
+  }
+  keys $data | each [key]{
+    echo (edit:styled $key":" blue) $data[$key]
+  }
+}
+
+# Uninstall a single package by removing its directory
+fn -uninstall-package [pkg]{
+  if (not (is-installed $pkg)) {
+    -error "Package "$pkg" is not installed."
+    return
+  }
+  dest = (dest $pkg)
+  -info "Removing package "$pkg
+  rm -rf $dest
+}
+
+######################################################################
+# Main user-facing functions
+
+# List installed packages
+fn installed {
+  e:ls $-lib-dir | each [dom]{
+    if ?(test -f (-domain-config-file $dom)) {
+      -read-domain-config $dom
+      lvl = $-domain-config[$dom][levels]
+      find $-lib-dir/$dom -type d -depth $lvl | each [pkg]{
```

Comment:
I would prefer that we do not rely on `find`, but it's OK for now.

Reply:
Any ideas about how to do this without `find`? I tried this, but it doesn't work (outputs the asterisks instead of expanding them), and it's not any shorter:
```
      put $-lib-dir/(repeat $lvl '*' | joins /) | each [pkg]{
        replaces $-lib-dir/ "" $pkg
      }
```
Another idea would be to `put $-lib-dir/**` and then filter by number of slashes (for the level) and by is-a-directory, but that seems even clunkier.

Reply:
Globbing a generated pattern is not supported at the moment, so an implementation without using `find` will need to traverse the filesystem recursively:

```
fn subdirs [dir level]{
  if (eq $level 0) {
    put $dir
  } else {
    put $dir/*/ | each [subdir]{ subdirs $subdir (- $level 1) }
}
```